### ISSUE TITLE
Missing Downloader implementations

### ISSUE THREAD

**Issue by aratare-jp** [TeamNewPipe/NewPipeExtractor#1159]
Hello there.

First of all, thanks a lot for this tool. I've been looking into using this for some of my personal projects and it has proven to be really great, despite the lack of tutorials/docs which makes things harder to understand for new people like me.

Anyway, onward to the actual question: I notice that, by default, there is no implementation class for `Downloader`, and it threw me off quite a bit when I wanted to parse a Youtube video URL with a `YoutubeStreamExtractor`.

I managed to find this in `NewPipe` and had to C&P it over and remove all of the Android-related code. So all is well now, but is there a reason for this?

**Comment by AudricV** [comment]
> I've been looking into using this for some of my personal projects and it has proven to be really great, despite the lack of tutorials/docs which makes things harder to understand for new people like me.

If you have some ideas or suggestions, or even if you want to contribute on that, free feel to open issues :)

> is there a reason for this?

Yes, the goal of the `Downloader` interface is to not depend on a specific network stack.

> I managed to find this in `NewPipe` and had to C&P it over and remove all of the Android-related code.

You could have looked at the [extractor test implementation](https://github.com/TeamNewPipe/NewPipeExtractor/blob/3402cdb66619202dcb7d6950f7c20fc78beef1fe/extractor/src/test/java/org/schabi/newpipe/downloader/DownloaderTestImpl.java) instead (note that the 429 HTTP response code handling is a very bad way to handle captcha responses at the moment, in the app and in the extractor).

Anyway, is the documentation of the `Downloader` class not enough to let you easily create an implementation?

**Comment by aratare-jp** [comment]
> If you have some ideas or suggestions, or even if you want to contribute on that, free feel to open issues :)
Yes I'm thinking about adding a new `How-to` section into the doc. This can be dedicated for different how-tos like fetching channels, video stream URLs, etc. Do let me know if you think this is helpful or not.

> Yes, the goal of the Downloader interface is to not depend on a specific network stack.
I understand that. My question was more about why there wasn't any "default" implementation of this so that people can use it out of the box. It actually took me a while to know that there wasn't one provided.

> is the documentation of the Downloader class not enough to let you easily create an implementation?
Create an implementation: yes. Create an implementation that works properly: not so easily.
For example, from the `DownloaderImpl` class within NewPipe, the `execute` function needs to set specific headers, handle ReCaptcha, set cookies, etc. in order to, my guess is, fetch stuff properly from Youtube. These little things are hard to know when one's required to create an implementation from scratch when they just want to make use of the library to fetch a streamable URL, for example.

**Comment by litetex** [comment]
Looks like this question was resolved, closing to clean stuff up.
